### PR TITLE
PWN-6092  - Selection mechanism for New Send Flow

### PR DESCRIPTION
**Pull request by 1sakov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-6092

## Description of Work

- Added initial logic for starting smart fee payer token selection
- Added transaction and account creation fee calculation
- Actual transaction sending
- Launching bottom transaction details bottom sheet

What was NOT done:

- Launching the fee payer screen. (https://github.com/p2p-org/p2p-wallet-android/pull/1014)
- Switching the fee payer from the token list but there is implemented fun for updating the Fee Payer Token. `fun updateFeePayerToken(feePayerToken: Token.Active)`

## Screenshots (Optional)

Screenshot #1 | Screenshot #2 | Screenshot #3
:------------:|:-------------:|:------------:
Screenshot_20221212_052956 | Screenshot_20221212_053030 | [Paste Pic 3]

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/newsend/SendFeeRelayerManager.kt [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+import org.p2p.wallet.send.model.FeePayerState
+import org.p2p.wallet.send.model.SearchResult
+import org.p2p.wallet.send.model.SendSolanaFee
+import timber.log.Timber
+import java.math.BigDecimal
+import java.math.BigInteger
+import kotlin.properties.Delegates
+import kotlinx.coroutines.CancellationException
+
+class SendFeeRelayerManager(
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/newsend/SendFeeRelayerManager.kt:25]: @gslevinkov Gleb GO GO
`Manager`

Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/newsend/SendFeeRelayerManager.kt:25]: not now i guess, send is needed asap :D

File: app/src/main/java/org/p2p/wallet/newsend/SendFeeRelayerManager.kt [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+                result = recipientAddress
+            )
+
+            if (feeRelayerFee == null) {
+                currentState = UpdateFee(solanaFee = null, feeLimitInfo = feeLimitInfo)
+                return
+            }
+
+            val inputAmount = tokenAmount.toLamports(sourceToken.decimals)
+            showFeeDetails(sourceToken, feeRelayerFee, feePayer, inputAmount, strategy)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/SendFeeRelayerManager.kt:86]: named arg

File: app/src/main/java/org/p2p/wallet/newsend/SendFeeRelayerManager.kt [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        val recipient = result.addressState.address
+
+        return try {
+            sendInteractor.calculateFeesForFeeRelayer(
+                feePayerToken = feePayerToken,
+                token = sourceToken,
+                recipient = recipient
+            )
+        } catch (e: CancellationException) {
+            Timber.w("Fee calculation is cancelled")
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/SendFeeRelayerManager.kt:111]: Timber.i

File: app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt [hunk 4]
```diff
@@ -179,15 +186,18 @@ class SendInteractor(
...
-        feeInSOL: BigInteger
-    ): BigInteger {
-        if (feePayerToken.isSOL) return feeInSOL
+        transactionFeeInSOL: BigInteger,
+        accountCreationFeeInSOL: BigInteger
+    ): FeeAmount {
+        if (feePayerToken.isSOL) return FeeAmount(
+            transaction = transactionFeeInSOL,
+            accountBalances = accountCreationFeeInSOL
+        )
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/interactor/SendInteractor.kt:195]:
```suggestion
        if (feePayerToken.isSOL) 
            return FeeAmount(
                transaction = transactionFeeInSOL,
                accountBalances = accountCreationFeeInSOL
            )
```

File: app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendTransactionsDetailsBottomSheet.kt [hunk 5]
```diff
@@ -25,21 +24,19 @@ class SendTransactionsDetailsBottomSheet : BaseDoneBottomSheet() {
...
 
     private lateinit var binding: DialogSendTransactionsDetailsBinding
 
-    private val state: SendTotal by args(ARG_SEND_STATE)
+    private val state: SendFeeTotal by args(ARG_SEND_STATE)
 
-    private val colorNight = getColor(R.color.text_night)
-    private val colorMountain = getColor(R.color.text_mountain)
-    private val colorMint = getColor(R.color.text_mint)
+    private val colorNight by lazy { getColor(R.color.text_night) }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/send/ui/dialogs/SendTransactionsDetailsBottomSheet.kt:37]: unsafe lazy is ok too

File: app/src/main/java/org/p2p/wallet/transaction/ui/NewProgressBottomSheet.kt [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+/**
+ * Bottom sheet dialog which shows current transaction's state
+ * The default state when it's launched is [TransactionState.Progress]
+ * */
+
+private const val EXTRA_DATA = "EXTRA_DATA"
+private const val EXTRA_TRANSACTION_ID = "EXTRA_TRANSACTION_ID"
+
+// TODO: change to NON-settling bottomSheet
+class NewProgressBottomSheet : NonDraggableBottomSheetDialogFragment() {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/transaction/ui/NewProgressBottomSheet.kt:33]: NewTransactionProgressBottomSheet?

**Review by gslevinkov — APPROVED** [review]

**Review by kabramovp2p — APPROVED** [review]

**Pull request merged**